### ISSUE TITLE
Train and deliver improved end-to-end offline WASM chess recognition

### ISSUE THREAD

**Issue by nino96** [nino96/chess-ocr#3]
## Outcome
Deliver a versioned offline recognizer that turns unfamiliar printed-diagram pages/selections into useful editable positions in the browser. A training run, larger dataset, successful export, or classifier-only improvement does not complete this issue.

Depends on merged #1 and #2's accepted first-learning-tranche interfaces/data records. Do not wait for full-corpus acquisition to begin useful learning; fresh qualification still requires its reviewed freeze and truth.

## Model and training decisions
- Start COCO-pretrained YOLOX-Nano for one-class board localization and native ImageNet-pretrained timm mobilenetv3_small_100.lamb_in1k for 13-class square recognition. Retain unchanged FENShot control. No reconstruction from inference ONNX needed; preserve original native base checkpoints locally with pinned manifests and verified reload.
- Classifier proposal: 96×96 RGB, checkpoint-appropriate normalization, new chess head. This differs from upstream resolution and must be measured; a single controlled resolution comparison is allowed only for demonstrated information loss, not an architecture sweep.
- Train head with backbone frozen first, then unfreeze later blocks at lower LR; deliberately control batch normalization statistics and regularization. Compare the resulting schedule against named failure hypotheses, not assumptions that fine-tuning guarantees transfer.
- Before meaningful optimization: audited image/label alignment, finite gradients, parameter update checks, tiny known-label fit, baseline equivalence where applicable, native/export/browser preprocessing agreement. Then CONTINUE into substantive training; a mechanics pilot is not the deliverable.
- Freeze data/lineage/rights/model/code hashes, seed(s), class/source sampling, real/synthetic mix, augmentation, optimizer/LR schedule, complete update schedule, checkpoint selection, confidence calibration and resource reservations. Record effective source repetition and per-class exposure. Reuse baseline outputs only for identical hashed inputs.
- Initial ~50/50 canonical real versus synthetic/degraded exposure is a prospective baseline; preserve clean and rare pieces, cap repetition of small sources. Do not silently replace the recipe after development outcomes.
- Declare adequate update counts and inspect full learning curves. Complete the schedule within the approved ledger; report budget-truncated runs as incomplete, not architecture failures. Save exact resumable native/optimizer/scheduler/RNG state and prove interruption recovery for the actual stochastic training path.
- Acquisition and training iterate within this issue and #2. A failed run prompts a bounded causal check: labels/rendering, preprocessing resolution, sampling/objective, forgetting, geometry or degradation. Do not default to more data, extra seeds or architecture search.

## Localization and complete path
- Train/evaluate full-page and loose-selection localization separately from oracle exact-crop classification. Refine INNER 8×8 grid excluding borders/coordinates; support declared small scan rotation/skew with corner geometry and uncertainty.
- Small diagrams require bounded overlapping page crops/scales; measure recall/latency tradeoff. Handle multiple boards, no-board pages, partial grids and unsupported inputs. No silent oracle geometry or choosing only successful detections.
- Report detector recall/false positives and normalized corner/grid error, classifier class/color/occupied confusion, and actual end-to-end exact positions. A good detector IoU alone does not establish correct tile alignment.
- Show board outlines and uncertain squares in #1's editable browser demo. User corrections are never overwritten by late results. Ship a versioned library/model manifest for chess-reader integration, not a second book reader.

## Promotion and qualification
Before new training, operationalize confidence/strata and minimum sample requirements with #2, including denominators and all-rejected behavior. Proposed retained gates:
- Paired >=5 percentage-point gain on the declared degraded/target development stratum against unchanged FENShot on IDENTICAL inputs; no loss of baseline-correct clean cases, occupied/class collapse, or increase in confidently wrong boards.
- Fresh source-held-out qualification after candidate freeze: >=95% reliable exact boards across all eligible inputs (misses/abstentions fail), >=99.5% accuracy among accepted/confident squares with coverage also reported, zero reliable-wrong boards observed. Report raw exact accuracy, source-group counts and uncertainty; zero observed errors is not a zero-risk guarantee.
- Frozen clean regression and negative/partial cases pass. Existing exposed evidence is diagnostic, not fresh qualification.
- Actual WASM CPU tests in Chromium/Firefox/WebKit, offline reload and zero unintended network, cancellation/timeouts/stale results/model integrity, and #1's measured size/memory/latency budgets pass. No arbitrary 2 MB cap and no WebGPU-only pass. Physical iPad remains explicitly unrun until authorized/named testing.

## Acceptance
- [ ] Candidate measurably improves the intended page/selection-to-position task, with paired gains/losses and localization/recognition errors separated.
- [ ] Native checkpoints, reproducible dataset/runtime manifests, full curves/confusion/results and resumable progress retained; safe summaries/recipes checked in, binaries distributed only through separately reviewed release mechanisms.
- [ ] Frozen qualification and clean-regression requirements pass or this issue remains incomplete with the exact blocker; no lowering gates after test exposure.
- [ ] Versioned browser library/demo works offline, uncertainty and edits are usable, real browser/runtime evidence recorded.
- [ ] Limits, licenses and all unrun physical-device gates are documented.

If offline quality or runtime cannot meet the declared outcome, report that evidence and activate #4's separately budgeted fallback; do not declare offline recognition solved.

**Comment by nino96** [comment]
## Corrective synthetic-bootstrap v2 checklist

This follows the parity correction recorded in #1. The current detector candidate is retained only as `legacy-bgr-div255-v1` diagnostic evidence; it is not qualified or a continuation source.

- [ ] Correct detector preprocessing: raw letterboxed RGB float32 `0..255`, pad `114` at the native/browser graph boundary; graph wrapper performs divide-by-255 plus ImageNet RGB mean/std normalization; training feeds the equivalent normalized RGB tensor.
- [ ] Replace fixed EMA decay with the pinned YOLOX ramp `0.9998 × (1 - exp(-updates / 2000))`; persist averaged weights and update count and prove exact stochastic resume.
- [ ] Persist checkpoint selection, final evaluation, and calibration before export; expose `trained`, `calibrating`, `exporting`, `export_failed`, and `complete` lifecycle states.
- [ ] Add export-only recovery: `pnpm run training -- export --run-root PATH --segment detector`, without repeating optimization, development evaluation, or calibration.
- [ ] Diagnose current live versus EMA detector weights at steps 1,500, 2,000, and 9,000.
- [ ] Enforce the bounded step-2,000 transition gate: stop before further GPU spend if live recall is below 5% after head-stage live recall reached at least 10%, or if loss/gradients become non-finite.
- [ ] Run detector-only v2 from the original COCO checkpoint, reusing the completed classifier checkpoint; do not continue the legacy detector.
- [ ] Complete confidence calibration, inner-grid refinement/rectification, source-held-out real-page evaluation, unchanged FENShot comparison, and native/ONNX/Chromium/Firefox/WebKit WASM gates before promotion.

Keep the first corrected run's existing 9,000-update schedule, seed, batch size, BN policy, and learning rates. No seed sweep, backbone-LR change, or new model family is authorized unless corrected live-model evidence still demonstrates unfreeze collapse. Reserve at most 600 preflight plus 5,400 detector GPU-seconds and retain the 24,000 CPU-second ceiling.

This remains a corrected synthetic bootstrap, not completion of #3.

**Comment by nino96** [comment]
Corrective synthetic detector bootstrap v2 is implementation-ready at commit `3ede91b`.

Frozen hypothesis: correcting the pinned checkpoint preprocessing and replacing fixed EMA with the pinned ramp will prevent the legacy detector's transfer/averaging failure without changing its seed, model family, schedule, batch size, BN policy, or learning rates.

- Recipe: `recipes/synthetic-bootstrap-v2.json`, SHA-256 `eb21bb1b0ba2115cbdd8865b540e2ca3e71ae7221f76936ab7e5acba5faa6120`
- Training implementation SHA-256: `97bedae2af489540f19b4ff59834f3ef0fd079f5f5bd5018d6a6a4c4bd670e19`
- Controller SHA-256: `3eaa32d87f546e3290c3700b1d23fb48ab2887af5c3a277832802f5818e54d93`
- Reused selected classifier checkpoint SHA-256: `acc6ebdafbd044483adeff29e30d853a6a5a83d9449c6499c047c2019b6f9f39`
- Changed inputs: raw letterboxed RGB float32 `0..255`, pad `114`; in-graph/training ImageNet normalization; EMA `0.9998 * (1 - exp(-updates / 2000))` with persisted update count.
- Command surface: detector-only `init`, then `start/status/stop`; export failure recovery is only `pnpm run training -- export --run-root PATH --segment detector`.
- Ceiling: 600 preflight plus 5,400 detector GPU-seconds; 24,000 CPU-seconds; 16 GiB output and 30% free-space floor.
- Stop: non-finite loss/gradient at any update, or at step 2,000 if live recall is below 5% after step-1,500 live recall reached at least 10%.
- Completion: all 9,000 updates plus persisted selection, final synthetic development evaluation, calibration, ONNX parity, then browser WASM checks. This remains synthetic diagnostic evidence only.

Pre-run gates passed: source/privacy/docs checks; 38 TypeScript tests; production build; 45 native/training tests; and 18 Chromium/Firefox/WebKit browser tests. No v2 GPU run had started when this record was posted.

**Comment by nino96** [comment]
Corrective synthetic detector v2 completed and the evaluation correction is ready for PR review on branch fix/synthetic-bootstrap-v2 (tip b6c1280). Run ec8d5a3fb66d74e5ec9dfafd36173988e8c22bb15b2d5b57cc51dbeb9e83c1ce completed all 9,000 updates and selected step 9,000. It used 3,225.380099814851 GPU-seconds; the cross-run deduplicated ledger is 12,496.457073617727 GPU-seconds in 12 unique attempt segments. The post-hoc CPU audit includes all 1,192 DEV pages (28 partial) and 708 CAL pages (12 partial), clamps per-IoU and aggregate AP, and preserves all 48 protected run files. Audit ID 7763cf7ecd9f80898df65a6e15d1255bcaded8d4a5689b4b83e94f22126f2eb4; report SHA-256 0d1fe1a8f23651c26923e6ad7461421984727299b1d2acb59102b6362d934f80. Corrected DEV AP50:95 is 0.9825175869014556; there were zero detections on 90 ordinary negatives but 93 on 28 partial pages. CAL had zero on 52 negatives but 41 on 12 partials. No nonzero-recall threshold meets the frozen no-valid-board false-positive limit, so corrected threshold is 1.0 with recall 0.0. One bounded diagnosis confirmed a partial synthetic input shows only five grid rows but enough board texture for YOLOX to localize; this is a downstream nine-line refinement/rejection blocker, not authority to relax calibration, retrain, sweep thresholds, or change models. Real development, shared grid refinement, browser WASM, and qualification remain incomplete.

**Comment by nino96** [comment]
## Real-development and qualification handoff

PR #11 delivered the frozen v2 browser/runtime integration. It did not establish
real-page promotion or qualification. Issue #3 remains dependent on issue #2's
resolved leakage blockers, prospective development membership, and accepted human
truth.

### Candidate and training decision

The first formal real-development comparison uses the existing frozen v2
candidate against unchanged FENShot on identical locked pages. No new training
run is authorized before that comparison.

After the formal result:

- if v2 passes, freeze it and proceed toward qualification without retraining;
- if partial-board failures arise from incomplete-grid acceptance, diagnose or
  fix deterministic refinement first;
- if evidence identifies a learned detector/classifier domain gap, a later
  TRAIN-only run may be proposed with a prospective recipe, hashes, resource
  reservation, stop criteria, and explicit owner approval; and
- a confidence-only failure may produce a separate owner-reviewed calibration
  proposal, not an automatic retraining run or threshold sweep.

Development and qualification pages never become training inputs.

### Partial-board treatment

A partial or unsupported diagram is a page with no valid complete board. It is
not relabeled as an ordinary negative, completed by inference, or given invented
64-square truth. Partial geometry remains separately recorded, but it supplies no
detector training target/image pair.

At runtime, the detector may emit a low-threshold region proposal; the nine-line
inner-grid refiner must reject an incomplete grid before classification. Formal
development includes a `negative-partial` stratum. Any returned board on such a
page is a false positive reported separately from ordinary negative-page false
positives; rejection or abstention is the expected outcome.

The synthetic audit already found a five-row partial grid that appeared
board-like to YOLOX. That evidence motivates the downstream rejection gate; it
does not authorize relaxed thresholds, fabricated labels, retraining, or a
threshold sweep.

### Formal development evaluator

Implement a separate `pnpm run development-evaluation --` surface with `plan`,
`status`, `stop`, `lock`, `run`, and `score`. It must require accepted human truth,
bind membership/provider/refiner/image/annotation hashes, reject qualification
inputs and stale or corrupt evidence, run frozen v2 and FENShot sequentially on
identical pages, enforce resource and cancellation boundaries, and clean up both
WASM sessions on every exit path.

Private raw evidence keeps page identities, positions, geometry, and model
results in ignored local storage. Only a separately privacy-reviewed aggregate
summary is suitable for an issue or results PR.

Score by model and stratum: missed and false boards, refinement failures,
normalized corner displacement, all-input exact boards, occupied/empty accuracy,
occupied-square class and color accuracy, confidently wrong squares and boards at
the frozen 0.99 threshold, abstention/coverage, cold/warm latency, and
FENShot-correct clean cases lost by v2.

The unchanged promotion gate requires:

- at least a five-percentage-point target/degraded exact-board improvement;
- no FENShot-correct clean page becoming wrong;
- v2 occupied/empty accuracy at least equal to FENShot;
- v2 occupied-square piece-class accuracy at least equal to FENShot; and
- no increase in confidently wrong boards.

Run each frozen model once after truth is locked. A failed gate permits one
bounded diagnosis, not automatic retraining, a new seed, a threshold sweep, a
backbone change, or another model family.

### Remote-browser operating note

When the demo is served through remote SSH, it still executes in the laptop's
browser. The browser file picker therefore sees the laptop filesystem. Copy these
ignored artifacts to a temporary laptop directory before selecting them:

```text
work/candidates/synthetic-bootstrap-v2-detector-v3.json
work/training/synthetic-bootstrap-v2-detector/classifier/selected.onnx
work/training/synthetic-bootstrap-v2-detector/detector/selected.onnx
```

The existing loader verifies their hashes and runs WASM on the laptop. It does not
upload the selected bytes. Do not expose arbitrary remote paths or directory
listings as a convenience mechanism.

Qualification begins only if v2 advances and its identity is frozen. Reuse the
formal evaluator with qualification-specific guards, then run the required
native, Node WASM, browser, offline reload, integrity, cancellation, stale-result,
latency, memory, and network gates without amending thresholds. Explicit owner
approval is required before every merge or issue closure.

**Comment by nino96** [comment]
## Fresh-session diagnostic handoff — 2026-09-09

The synthetic-bootstrap v2 detector and classifier remain frozen diagnostic baselines and are rejected for promotion in their current form. Unchanged FENShot remains the control. Private diagnostic inputs and raw results remain local and are deliberately omitted from GitHub.

The work is now separated into two bounded, independently reviewable causes:

- #18 owns automatic localization/refinement: edge-touching-grid incompatibility, internal false-grid acceptance, corner canonicalization, proposal threshold/NMS/cap visibility, and raw per-proposal refinement evidence.
- #19 owns classifier diagnosis: exact manual-grid tile/tensor/logit parity across training/native/ONNX/browser paths, followed by a materially revised training proposal if parity passes. The unchanged synthetic-only recipe must not be rerun.

Issue #2 owns the required source-diverse real TRAIN collection and expanded reviewed printed-diagram artwork. Exposed diagnostic and development inputs remain excluded from training unless separately reclassified with explicit owner authority and leakage review; the current private diagnostics are not authorized for training.

PR #17 is open with all checks passing. It delivers the usable manual four-corner workflow, side-by-side source/reference editing, exact `.PNBRQKpnbrqk` keyboard entry, and visibility for every accepted grid, rectified board, prediction, and warning. It does not fix model behavior or expose raw YOLOX proposals.

Recommended next assignment: complete #19's bounded parity audit first because it decides whether the next classifier work is an implementation correction or revised training. Then prepare the reviewable training freeze through #3 and #2. #18 can be assigned separately because it keeps classifier weights and preprocessing fixed while correcting geometry. Work on one assigned issue/branch at a time and do not launch training from uncommitted code.

No GPU training job, architecture change, qualification run, cloud use, private upload, or publication is authorized by this handoff.
